### ISSUE TITLE
Flexible keymap (hotkeys / shortcut)

### ISSUE THREAD

**Issue by iamstarkov** [zeit/hyper#872]
the important key to achieve it in flexible and reliable way is to have two things:
- list of possible events to `rpc.emit` <sup>1</sup>
- one place to bind keys to those events

<sup>1</sup> Note: `rpc` bus should be the only way to broadcast key events).

Initial array of events can be collected from [`lib/index.js#L38-L129`](https://github.com/zeit/hyper/blob/aa285a8bc662f7fae1f9f265f326c87c2de88420/lib/index.js#L38-L129): like `split request vertical` or  `split request horizontal`, etc.

Then in config we can do smth like this: 

``` js
keymap: {
 "CmdOrCtrl+Alt+Left": "prev pane req",
}
```

where keys are electron [accelerators](https://github.com/electron/electron/blob/master/docs/api/accelerator.md) and values are those events for rpc to emit/receive.

After that, we need to modify `attachKeyListeners()` [lib/containers/hyper.js#L38-L76](https://github.com/zeit/hyper/blob/master/lib/containers/hyper.js#L38-L76) to not bind hardcoded keys, but instead to pick up keys and events from keymap:

``` js
// mousetrapify needs to be more advanced, for sure, but this is enough for now
const mousetrapify = key => key.replace('CmdOrCtrl', 'mod').toLowerCase();
Object.keys(this.config.keymap).map(item => {
  const accelerator = item;
  const event = this.config.keys[key];
  keys.bind(
    mousetrapify(accelerator),
    () => {
      if (focusedWindow) {
          focusedWindow.rpc.emit(event);
      }
    }
 );
})
```

Also, Hyper needs to verify that this.config is merged object similar to`Object.assign({}, defaultConfig, ...[pluginsAndThemesConfigs], userConfig)`

_Thats it._

---

One interesting benefit of proposed approach is that _you can update menu accelerators very easy and smooth_. just needs to write a function which will lookup desired accelerator by key event relevant to that menu item. <sup>2</sup>, <sup>3</sup>
-   <sup>2</sup> because right now [keeping hotkeys in a menu relevant is a fragile mess](https://github.com/iamstarkov/hyper-panes-iterm2-hotkeys/blob/master/index.js#L27-L43)
-  <sup>3</sup> findAccelerator is pretty simple function:
  
  ``` js
  const findAccelerator = event => Object.keys(this.config)
    .map(key => { key, value: this.config[key] })
    .find({ value } => value === event)
    .key
  ```
### Result
- widely commons keymap sets can be distributed via plugins
- always relevant hotkeys in menu to briefly lookup forgotten keybindings
- with list of events and keymap to them it will be possible to "view all keyboard shortcuts" (fix #814)
- _user gets fully customizable key bindings 🎉_ (fix #251 and #657)

what do you think?

**Comment by iamstarkov** [comment]
i want to discuss this approach before i will start implementation, so any feedback is welcomed: ideas, questions, concerns, problems, etc

**Comment by iamstarkov** [comment]
@leo @CodeTheory @ekmartin @rauchg what do you think?

**Comment by ekmartin** [comment]
I'm not sure if it's a good idea to bind keybindings to rpc directly. Far from all combinations have an rpc event connected to them at the moment, so not sure if that's a good direction to go.

Why not have a separate keymap file that maps a string key to a combination? I.e. `'split-vertical': 'Cmd+D'`. That way the keymap config could be used both in the menu decorator and in the [hyper container](https://github.com/zeit/hyper/blob/master/lib/containers/hyper.js#L47-L74).
We could then expose the keymap to plugins in the same way we do with a regular config.

Example would be: `keys.bind(keyMap['move-right'], moveRight)`.

**Comment by iamstarkov** [comment]
> I'm not sure if it's a good idea to bind keybindings to rpc directly.

maybe its not a good idea, but rpc as a central piece of code for key bindings — is the best point i could think of. To be honest, in the end it can be not a rpc, but any central broadcast communication api. It just happens, that rpc is what we have now.

---

> Far from all combinations have an rpc event connected to them at the moment, so not sure if that's a good direction to go.

We can add them later, once proposed changes will be in place, it should be easy.

---

> Why not have a separate keymap file that maps a string key to a combination?

main reason against two files over one, is simplicity: _harder to show/sync your settings with two configuration files_ You may think that configuration file with a whole keymap will be too hard to read, though to cure this problem we need to think careful about sane default keymap.

---

> I.e. 'split-vertical': 'Cmd+D'. That way the keymap config could be used both in the menu decorator and in the hyper container.

Exactly! I'm all for that as i mentioned in proposal.

Need to note, that `keybinding: rpcEvent` have to be reversed, because relation is not one keybinding to multiply rpcEvents, but quite opposite — _one rpcEvent to many keybindings_. And hey, its how its done in Atom, which i cant blame for blame for bad keymap decisions.

---

> We could then expose the keymap to plugins in the same way we do with a regular config.

Yep!

---

> Example would be: keys.bind(keyMap['move-right'], moveRight).

Kinda, as far as pairs are revered you have to do smth like this:

```js
Object.keys(this.config.keymap).map(accelerator => {
      const event = this.config.keys[key];
      keys.bind(
        mousetrapify(accelerator),
        () => { rpc.emit(event); } // kinda
     );
    })
```

**Comment by iamstarkov** [comment]
tldr:

* one file is simpler than two
* for this feature to work we need broadcast mechanism, lets start with `rpc` for now
* pair should be `keybinding: event`

**Comment by jwhitley** [comment]
I'll argue for an approach to keybindings that guarantees sensible rebinding and/or unbinding.  A concrete example:

When using iTerm2, I rebind the ⌘1 through ⌘9 keys globally (across all profiles).  Instead of their normal function of switching the first nine iTerm2 tabs, they instead are bound to emit `<prefix>-<num>` which switches between tmux windows.  I.e. I use tmux so heavily that I prefer the normal Command bindings to treat tmux windows like tabs.

Another frustration of some keybinding systems (vim, I'm lookin' at you) is that not all bindings are first-class and/or introspectable.  I.e. there's no command or API that can ask such questions as:

1.  What are *all* the currently active keybindings?
2.  What key(s) are bound to operation `<foo>`?
3.  What operation(s) are bound to key-sequence `<bar>`?
4.  (Bonus points) Where did a given binding come from?  (e.g. user config, plugin `baz`, etc.)

These operations aid both config debugging and discoverability.

**Comment by iamstarkov** [comment]
@jwhitley After and if proposed change will land in master. All keybindings will be declared in one `keymap` object, and it will be possible to redefine them in your `config.keymap`

**Comment by iamstarkov** [comment]
bad news everyone, Hyper has two rpc, one for Electron and one for Renderer processes, and you cant share them. problem here is that key bindings in `./lib/containers/hyper.js` can be used to emit event only to to Renderer's rpc, while actual rpc responsible for hotkeys is Electron's one

**Comment by iamstarkov** [comment]
on the other not, there is another way

**Comment by iamstarkov** [comment]
This PR #925 is another approach, i was talking about

**Comment by iamstarkov** [comment]
@rauchg can i ask you to move labels from #925 PR into this original issue? `Priority: High`, `Status: Community feedback wanted`, `Status: In Progress`. And share your thoughts about last label

**Comment by ppot** [comment]
@iamstarkov I switched the label

**Comment by ppot** [comment]
Plz add request inside issue: ppot#11

**Comment by andrroy** [comment]
Referring to the label "In progress", is this issue actively being worked on? All i could find was @iamstarkov closed PR.

**Comment by iamstarkov** [comment]
@andrroy look at  https://github.com/ppot/hyper/issues/11

**Comment by iamstarkov** [comment]
and yes it is in progress

**Comment by iamstarkov** [comment]
here https://github.com/zeit/hyper/pull/1509

**Comment by ppot** [comment]
close in #1876
